explainer-datamodel.md

```diff
@@ -0,0 +1,413 @@
+mode objects, reusing the web platform's existing line formatting and wrapping engine along with 
+support for CSS, while also enabling authors to apply the objects to a Canvas how and when they
+choose. Alternately, authors may not desire to render the objects to a Canvas at all, and instead
+can run a given layout over them to obtain metrics useful in other scenarios.
+
+In this explainer, a new formatted text data model is presented (the "retained" mode objects) which
+is the minimal necessary objects to allow CSS-styled text runs to be collected in a sequence
+for layout processing.
+
+### Principles
+* An imperative JavaScript-friendly text representation.
+* Scope to the needs of inline text layout.
+* Leverage CSS as the universal text layout system for the web (even in an imperative model) which also provides for future extensibility.
+* Avoid multi-level hierarchal text structures (which must be linearly iterated for rendering anyway)
+* Object-centric model (versus range+indexes) to improve encapsulation and avoid problems with overlapping formatting.
+
+### Primary Objects
+
+The data model for formatted text is quite simple and involves only two objects: `FormattedText`
+is a container that orders a list of `FormattedTextRun` objects that contain the actual text and
+formatting information. (The `FormattedText` container can also have formatting applied that may
+inherit down to the `FormattedTextRun`s.)
+
+To maximize authoring convenience, we are proposing using the 
+[ObserverableArray pattern](https://heycam.github.io/webidl/#idl-observable-array) to allow authors
+to manage the contained array exactly like a JavaScript Array.
+
+<img src="explainerresources/formatted-text-om.png" alt="FormattedText object holds an array of FormattedTextRun objects through a property called 'textruns'." align="center"/>
+
+## Creating FormattedText and FormattedTextRun objects
+
+`FormattedText` and `FormattedTextRun` objects are created using the 'new' pattern:
+
+```js
+let text = new FormattedText();
+let textrun = new FormattedTextRun();
+```
+
+Attach `FormattedTextRun` objects to their `FormattedText` object's `textruns` property (an Array):
+
+```js
+text.textruns.push( textrun );
+```
+
+Add text to the `FormattedTextRun` object:
+
+```js
+textrun.text = "hello";
+```
+
+To simplify creation of a `FormattedText` object and its textruns, the constructors also supports
+various overloads:
+
+```js
+let text = new FormattedText("hello", " world");
+// equivalent to:
+let text = new FormattedText();
+text.textruns.push( new FormattedTextRun( {text: "hello"} ) );
+text.textruns.push( new FormattedTextRun( " world" ) );
+// make a copy of all the text runs from another FormattedText object to be included in a new one:
+let text2 = new FormattedText( text.textruns );
+// copy everything from another FormattedText object and its text runs
+let text3 = new FormattedText( text2 );
+```
+
+## Common Styles
```

Comment:
I dropped the common styles, to simplify how cascading and style computation implementation. Wanted to hear your thoughts on dropping it ? I thought it was more consistent.

Reply:
I completely agree with the consistency aspect. I'll cleanup other examples to remove other traces...

Reply:
After dropping `font` and `color`, I'm actually going to add `style` as something that can be set at construction time as a shortcut to simplify usage of the API. It's more generic and feeds nicely into `setStyle`, so that you can still create text with style at construction time with one call ;)

Reply:
See https://github.com/WICG/canvas-formatted-text/commit/1ac4886ac55b5d60cdb4f6391d218ae28d5e4e73